packages/theme/src/cli/utilities/repl/evaluater.ts

```diff
@@ -0,0 +1,199 @@
+}
+
+async function evalContext(config: EvaluationConfig) {
+  outputDebug(`Evaluating context - ${config.snippet}`)
+
+  const json = `{ "type": "context", "value": "{% ${config.snippet.replace(/"/g, '\\"')} %}" }`
+  const {status, text} = await makeRequest({...config, snippet: json})
+
+  if (successfulRequest(status, text)) {
+    config.replSession.push(JSON.parse(json))
+  }
+}
+
+async function evalAssignmentContext(config: EvaluationConfig) {
+  outputDebug(`Evaluating assignment context - ${config.snippet}`)
+
+  if (isSmartAssignment(config.snippet)) {
+    config.snippet = `assign ${config.snippet}`
+    outputInfo(outputContent`${outputToken.gray(`> ${config.snippet}`)}`)
+    return evalContext(config)
+  }
+}
+
+async function evalSyntaxError(config: EvaluationConfig) {
+  outputDebug(`Evaluating syntax error - ${config.snippet}`)
+
+  let body = ''
+  if (!isStandardAssignment(config.snippet)) {
+    const {text} = await makeRequest({...config, snippet: `{{ ${config.snippet} }}`})
+    body = text
+  }
+
+  if (!hasLiquidError(body)) {
+    const {text} = await makeRequest({...config, snippet: `{% ${config.snippet} %}`})
+    body = text
+  }
+
+  if (hasLiquidError(body)) {
+    const error = body.replace(/ \(snippets\/eval line \d+\)/, '')
+    printSyntaxError(config.snippet, error)
+  }
+}
+
+function printSyntaxError(snippet: string, error: string) {
+  if (error.includes('Unknown tag')) {
+    outputInfo(outputContent`${outputToken.errorText(`Unknown object, property, tag, or filter: '${snippet}'`)}`)
+    return
+  }
+
+  const match = stripHTMLContent(error)
+  if (match && match[1]) {
+    outputInfo(outputContent`${outputToken.errorText(match[1])}`)
+  }
+}
+
+async function makeRequest(config: EvaluationConfig): Promise<{text: string; status: number}> {
+  const requestBody = buildRequestBody(config)
+  const response = await sendRenderRequest(config, requestBody)
+  return {text: await response.text(), status: response.status}
+}
+
+function buildRequestBody(config: EvaluationConfig): string {
+  const items = [...config.replSession.map((item) => JSON.stringify(item)), config.snippet]
+  return `[${items.join(',').replace(/\\"/g, '"')}]`
+}
+
+async function sendRenderRequest(config: EvaluationConfig, requestBody: string) {
+  const response = await render(config.themeSession, {
+    path: config.url,
+    query: [],
+    themeId: config.themeId,
+    cookies: '',
+    sectionId: 'announcement-bar',
+    headers: {},
+    replaceTemplates: {
+      'sections/announcement-bar.liquid': `\n{% render 'eval' %}`,
+      'snippets/eval.liquid': requestBody,
+    },
+  })
+
+  if (isExpiredSession(response.status)) {
+    expiredSessionError()
+  }
+
+  if (isTooManyRequests(response.status)) {
+    tooManyRequestsError()
+  }
+
+  const serverTiming = response.headers.get('server-timing')
+  if (isResourceNotFound(serverTiming || '')) {
+    notFoundError()
+  }
```

Comment:
I believe we could make each method hold the logic to identify the type of response, something like this:

```suggestion
  if (isExpiredSession(response)) {
    expiredSessionError()
  }

  if (isTooManyRequests(response)) {
    tooManyRequestsError()
  }

  if (isResourceNotFound(response)) {
    notFoundError()
  }
```

To make that happen, we need to update the other methods to:
```ts
// import {Response} from '@shopify/cli-kit/node/http'
function isExpiredSession(response: Response): boolean {
  return response.status === 401 || response.status === 403
}

function isTooManyRequests(response: Response): boolean {
  return response.status === 430 || response.status === 429
}

function isResourceNotFound(response: Response): boolean {
  // We don't look for the status code here because the Section Rendering API returns 200 even on unknown paths.
  return (response.headers.get('server-timing') ?? '').includes('pageType;desc="404"')
}
```

And fix the `Response` type at `packages/cli-kit/src/public/node/http.ts` to use `Awaited`:

![image](https://github.com/user-attachments/assets/2c5ee791-baf7-44e7-a59e-f75333b7ca19)